winton_kafka_streams/processor/_stream_thread.py

```diff
@@ -189,28 +254,57 @@ def run(self):
             self.consumer.subscribe(self.topics, on_assign=self.on_assign, on_revoke=self.on_revoke)
 
             while self.still_running():
-                record = self.consumer.poll(0.1)
-                if record is None:
-                    continue
-                elif not record.error():
-                    log.debug('Received message: %s', record.value().decode('utf-8'))
-                    self.processAndPunctuate(record)
-                elif record.error().code() == KafkaError._PARTITION_EOF:
-                    continue
-                elif record.error():
-                    log.error('Record error received: %s', record.error())
+                records = self.poll_requests(0.1)
+                if records:
+                    log.debug(f'Processing {len(records)} record(s)')
+                    self.add_records_to_tasks(records)
+                    self.process_and_punctuate()
 
             log.debug('Ending stream thread...')
         finally:
             self.commitAll()
             self.shutdown()
 
-    def processAndPunctuate(self, record):
-        task = self.tasks[record.partition()]
-        task.add_records([record])
-        task.process()
-        if task.needCommit:
-            self.commit(task)
+    def poll_requests(self, poll_timeout):
+        """ Get the next batch of records """
+
+        # The current python kafka client gives us messages one by one,
+        # but for better throughput we want to process many records at once.
+        # Keep polling until we get no more records out.
+        records = []
+        record = self.consumer.poll(poll_timeout)
+        while record is not None:
```

Comment:
This would be something that would be application specific. Messages could be very large and we may not want to take so many into the queue or we may want to trade latency with more dependable outputs timing for example. As I said not perhaps a problem immediately.